app/Models/Foundation/Summit/Events/Presentations/Presentation.php

```diff
@@ -339,31 +334,84 @@ public function getClassName()
      */
     public function getSpeakers()
     {
-        return $this->speakers;
+        return $this->speakers->map(function ($entity) {
+            return $entity->getSpeaker();
+        });
+    }
+
+    /**
+     * @return int|null
+     */
+    public function getSpeakerAssignmentsMaxOrder(): int
+    {
+        $criteria = Criteria::create();
+        $criteria->orderBy(['order' => 'DESC']);
+        $res = $this->speakers->matching($criteria)->first();
+        return $res === false ? 0 : $res->getOrder();
     }
 
     /**
      * @param PresentationSpeaker $speaker
      */
     public function addSpeaker(PresentationSpeaker $speaker)
     {
-        if ($this->speakers->contains($speaker)) return;
-        $this->speakers->add($speaker);
-        $speaker->addPresentation($this);
+        $speakers = $this->getSpeakers();
+        if ($speakers->contains($speaker)) return;
+        $order = $this->getSpeakerAssignmentsMaxOrder();
+        $speaker_assignment = new PresentationSpeakerAssignment($this, $speaker, $order + 1);
+        $this->speakers->add($speaker_assignment);
     }
 
     public function clearSpeakers()
     {
         $this->speakers->clear();
     }
 
+    /**
+     * @param PresentationSpeaker $speaker
+     */
+    public function removeSpeaker(PresentationSpeaker $speaker)
+    {
+        $criteria = Criteria::create();
+        $criteria->where(Criteria::expr()->eq('speaker', $speaker));
+        $speaker_assignment = $this->speakers->matching($criteria)->first();
+        if ($speaker_assignment === false) return;
+        $this->speakers->removeElement($speaker_assignment);
```

Comment:
do we need to clear up speaker_assignment.speaker and speaker_assignment.presentation here ?

Reply:
I don't think we need that clearing. This is meant just to remove the assignment.